Service status API documentation refactor (#14528)
apache/druid: docs/api-reference/service-status-api.md

        },
        {
            "name": "org.apache.druid.msq.guice.MSQIndexingModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.MSQDurableStorageModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.MSQServiceClientModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.MSQSqlModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.SqlTaskModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        }
    ],
    "memory": {
        "maxMemory": 268435456,
        "totalMemory": 268435456,
        "freeMemory": 139060688,
        "usedMemory": 129374768,
        "directMemory": 134217728
    }
  }
  ```
</details>

### Get process health

#### URL

<code class="getAPI">GET</code> `/status/health`

> Retrieves the health of the Druid service. If online, it will always return a JSON object with the boolean `true` value, indicating that the service can receive API calls. This endpoint is suitable for automated health checks.

Author: endpoint only returns true or nothing.

    "os.arch": "aarch64",
    "java.vm.specification.name": "Java Virtual Machine Specification",
    "druid.node.type": "router",
    "java.awt.printerjob": "sun.lwawt.macosx.CPrinterJob",
    "sun.os.patch.level": "unknown",
    "java.util.logging.manager": "org.apache.logging.log4j.jul.LogManager",
    "java.library.path": "/Users/genericUser/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:.",
    "java.vendor": "Homebrew",
    "java.vm.info": "mixed mode",
    "java.vm.version": "11.0.19+0",
    "druid.emitter": "noop",
    "sun.io.unicode.encoding": "UnicodeBig",
    "druid.storage.type": "local",
    "druid.expressions.useStrictBooleans": "true",
    "java.class.version": "55.0",
    "socksNonProxyHosts": "local|*.local|169.254/16|*.169.254/16",
    "druid.server.hiddenProperties": "[\"druid.s3.accessKey\",\"druid.s3.secretKey\",\"druid.metadata.storage.connector.password\", \"password\", \"key\", \"token\", \"pwd\"]"
}
```
</details>

### Get node discovery status and cluster integration confirmation

#### URL
<code class="getAPI">GET</code> `/status/selfDiscovered/status`

> Retrieves a JSON map of the form `{"selfDiscovered": true/false}`, indicating whether the node has received a confirmation from the central node discovery mechanism (currently ZooKeeper) of the Druid cluster that the node has been added to the

Reviewer: Similar comment to line 216. Is this a check we need to perform against other services/ports than just the router?

Author: this is a per process/node query. so you only run it once per node and it can be queried for each service

cluster. 

> It is recommended to only consider a Druid node "healthy" or "ready" in automated deployment/container management systems when it returns `{"selfDiscovered": true}` from this endpoint. Nodes experiencing network issues may become isolated and should not be considered "healthy." Nodes utilizing Zookeeper segment discovery may be unusable until the Zookeeper client is fully initialized and receives data from the Zookeeper cluster. The presence of `{"selfDiscovered": true}` indicates that the node's Zookeeper client has started receiving data, enabling timely discovery of segments and other nodes.

Reviewer: ```suggestion
Only consider a Druid node "healthy" or "ready" in automated deployment/container management systems when this endpoint returns `{"selfDiscovered": true}`. Nodes experiencing network issues may become isolated and are not healthy.

For nodes that use Zookeeper segment discovery, a response of `{"selfDiscovered": true}` indicates that the node's Zookeeper client has started receiving data from the Zookeeper cluster, enabling timely discovery of segments and other nodes.
```
"It is recommended" is both passive and unnecessary. Reworded

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Node was successfully added to the cluster*  
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell
curl "{domain}/status/selfDiscovered/status"
```
<!--HTTP-->
```http
GET /status/selfDiscovered/status HTTP/1.1
Host: {domain}
```

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>
  ```json
  {
    "selfDiscovered": true
  }
  ```
</details>

### Get node self-discovery status

#### URL
<code class="getAPI">GET</code> `/status/selfDiscovered`

> Retrieves a status code to indicate if a node discovered itself within the Druid cluster. This is similar to the `status/selfDiscovered/status` endpoint but relies on HTTP status codes alone. This is useful for certain monitoring checks such as AWS load balancer health checks that are unable to examine the response body.

Reviewer: ```suggestion
Returns an HTTP status code to indicate node discovery within the Druid cluster. This endpoint is similar to the `status/selfDiscovered/status` endpoint, but relies on HTTP status codes alone.

Use this endpoint for monitoring checks that are unable to examine the response body. For example, AWS load balancer health checks.
```

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell
curl "{domain}/status/selfDiscovered"
```
<!--HTTP-->
```http
GET /status/selfDiscovered HTTP/1.1
Host: {domain}
```

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
A successful response to this endpoint results in an empty response body.

## Coordinator

### Get leader address

#### URL

<code class="getAPI">GET</code> `/druid/coordinator/v1/leader`

Retrieves the address of the current leader Coordinator of the cluster.

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully retrieved leader Coordinator address*  
<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell
curl "{domain}/druid/coordinator/v1/leader"
```
<!--HTTP-->
```http
GET /druid/coordinator/v1/leader HTTP/1.1
Host: {domain}
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>
  ```json
  http://localhost:8081
  ```
</details>

### Get leader status

#### URL
<code class="getAPI">GET</code> `/druid/coordinator/v1/isLeader`

> Retrieves a JSON object with a `leader` key. The value can be `true` or `false`, indicating if this server is the current leader Coordinator of the cluster. This is suitable for use as a load balancer status check if you only want the active leader to be considered in-service at the load balancer.

Reviewer: Would you need to run this against all the individual Coordinator endpoints?

Author: I think with the /leader/ endpoint, there isn't a need to run it against all coordinators since /leader/ will check all of them and /isLeader/ is for a particular one.

Author: No, this is a per process/node query.

> Retrieves a JSON object with a `leader` key. The value can be `true` or `false`, indicating if this server is the current leader Coordinator of the cluster. This is suitable for use as a load balancer status check if you only want the active leader to be considered in-service at the load balancer.

Reviewer: ```suggestion
Retrieves a JSON object with a `leader` key. 

Returns `true` if this server is the current leader Coordinator of the cluster.

Use this endpoint as a load balancer status check when you only want the active leader to be considered in-service at the load balancer.
```

GET /druid/coordinator/v1/isLeader HTTP/1.1
Host: {domain}
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>
  ```json
  {
    "leader": true
  }
  ```
</details>

<a name="coordinator-segment-loading"></a>

## Overlord

### Get leader address

#### URL

<code class="getAPI">GET</code> `/druid/indexer/v1/leader`

Retrieves the address of the current leader Overlord of the cluster. In a cluster of multiple Overlords, only one Overlord assumes the leading role, while the remaining Overlords remain on standby.

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully retrieved leader Overlord address*  
<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell
curl "{domain}/druid/indexer/v1/leader"
```
<!--HTTP-->
```http
GET /druid/indexer/v1/leader HTTP/1.1
Host: {domain}
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>

  ```json
  http://localhost:8081
  ```

</details>

### Get leader status

#### URL
<code class="getAPI">GET</code> `/druid/indexer/v1/isLeader`

> Retrieves a JSON object with a `leader` property. The value can be `true` or `false`, indicating if this server is the current leader Overlord of the cluster. This is suitable for use as a load balancer status check if you only want the active leader to be considered in-service at the load balancer.

Reviewer: Similar to 478. Do you need to run this against all the overlord nodes, not just the router?

Also recommend similar stylistic updates

  ~ under the License.
  -->

This document describes the API endpoints to retrieve service status, cluster information for Apache Druid.

In this document, `http://<SERVICE_IP>:<SERVICE_PORT>` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://<ROUTER_IP>:<ROUTER_PORT>` with `http://localhost:8888`.

## Common

All services support the following endpoints.

> Each endpoint can be used with the ports for each type of service.  The following table contains port addresses for a local configuration:

Reviewer: ```suggestion
You can use each endpoint with the ports for each type of service.  The following table contains port addresses for a local configuration:
```

---

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell
curl "http://<ROUTER_IP>:<ROUTER_PORT>/druid/coordinator/v1/leader"
```
<!--HTTP-->
```http
GET /druid/coordinator/v1/leader HTTP/1.1
Host: http://<ROUTER_IP>:<ROUTER_PORT>
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>
  ```json
  http://localhost:8081
  ```
</details>

### Get Coordinator leader status

#### URL
<code class="getAPI">GET</code> `/druid/coordinator/v1/isLeader`

Retrieves a JSON object with a `leader` key. Returns `true` if this server is the current leader Coordinator of the cluster. To get the individual address of the leader Coordinator node, see the [leader endpoint](#get-coordinator-leader-address).

Use this endpoint as a load balancer status check when you only want the active leader to be considered in-service at the load balancer. /coo

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Current server is the leader*  

<!--404 NOT FOUND->
<br/>
*Current server is not the leader*  

<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell

> curl "http://<COORDINATOR_IP>:<COORDINATOR_IP>/druid/coordinator/v1/isLeader"
> ```
> <!--HTTP-->
> ```http
> GET /druid/coordinator/v1/isLeader HTTP/1.1
> Host: http://<COORDINATOR_IP>:<COORDINATOR_IP>

Reviewer: ```suggestion
curl "http://<COORDINATOR_IP>:<COORDINATOR_PORT>/druid/coordinator/v1/isLeader"
```
<!--HTTP-->
```http
GET /druid/coordinator/v1/isLeader HTTP/1.1
Host: http://<COORDINATOR_IP>:<COORDINATOR_PORT>
```
typo ip:port

  }
  ```

</details>

### Enable MiddleManager

#### URL

<code class="postAPI">POST</code> `/druid/worker/v1/enable`

Enables a MiddleManager, allowing it to accept new tasks again if it was previously disabled. Returns a JSON  object
keyed by the combined `druid.host` and `druid.port`.

#### Responses
<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Successfully enabled MiddleManager*  

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell
curl "http://<MIDDLEMANAGER_IP>:<MIDDLEMANAGER_PORT>/druid/worker/v1/enable"
```
<!--HTTP-->
```http
POST /druid/worker/v1/enable HTTP/1.1
Host: http://<MIDDLEMANAGER_IP>:<MIDDLEMANAGER_PORT>
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>
  
  ```json
  {
    "localhost:8091":"enabled"
  }
  ```

</details>

## Historical  

### Get segment load status

#### URL

<code class="getAPI">GET</code> `/druid/historical/v1/loadstatus`

Retrieves a JSON object of the form `{"cacheInitialized":<value>}`, where value is either `true` or `false` indicating if all segments in the local cache have been loaded. 

> Use this endpoint to know when a Historical process is ready to be queried after a restart.

Reviewer: ```suggestion
Use this endpoint to know when a Historical process is ready to accept queries after a restart.
```
nit passive

    "cacheInitialized": true
  }
  ```
</details>

### Get segment readiness

#### URL
<code class="getAPI">GET</code> `/druid/historical/v1/readiness`

Retrieves a status code to indicate if all segments in the local cache have been loaded. Similar to `/druid/historical/v1/loadstatus`, but instead of returning JSON with a flag, it returns status codes.

#### Responses
<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Segments in local cache successfully loaded*  
<!--503 SERVICE UNAVAILABLE-->
<br/>
*Segments in local cache have not been loaded*  

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell
curl "http://<HISTORICAL_IP>:<HISTORICAL_PORT>/druid/historical/v1/readiness"
```
<!--HTTP-->
```http
GET /druid/historical/v1/readiness HTTP/1.1
Host: http://<HISTORICAL_IP>:<HISTORICAL_PORT>
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
A successful response to this endpoint results in an empty response body.

## Load Status

### Get Broker query load status

#### URL

<code class="getAPI">GET</code> `/druid/broker/v1/loadstatus`

Retrieves a flag indicating if the Broker knows about all segments in the cluster. 

> Use this endpoint to know when a Broker service is ready to be queried after a restart.

Reviewer: ```suggestion
Use this endpoint to know when a Broker service is ready to accept queries after a restart.
```
nit passive

#### Responses
<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->
<br/>
*Segments successfully loaded*  

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->
<!--cURL-->
```shell
curl "http://<BROKER_IP>:<BROKER_PORT>/druid/broker/v1/loadstatus"
```
<!--HTTP-->
```http
GET /druid/broker/v1/loadstatus HTTP/1.1
Host: http://<BROKER_IP>:<BROKER_PORT>
```
<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response
<details>
  <summary>Click to show sample response</summary>
  ```json
  {
    "inventoryInitialized": true
  }
  ```
</details>

### Get Broker query readiness

#### URL

<code class="getAPI">GET</code> `/druid/broker/v1/readiness`

> Retrieves a status code indicating if the Broker knows about all segments in the cluster and is ready to be queried after a restart. Similar to `/druid/broker/v1/loadstatus`, but instead of returning a JSON, it returns status codes.

Reviewer: ```suggestion
Retrieves a status code indicating if the Broker knows about all segments in the cluster and is ready to accept queries after a restart. Similar to `/druid/broker/v1/loadstatus`, but instead of returning a JSON, it returns status codes.
```
nit passive. "status code indicating if the Broker knows about all segments" reads awkwardly, but I can't think of a better way to reword it.

  ~ under the License.
  -->

This document describes the API endpoints to retrieve service status, cluster information for Apache Druid.

In this document, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://ROUTER_IP:ROUTER_PORT` with `http://localhost:8888`.

## Common

All services support the following endpoints.

You can use each endpoint with the ports for each type of service.  The following table contains port addresses for a local configuration:

> |Service|Port address|
> | ----------- | ----------- |
> | Coordinator|8081|
> | Overlord   |8081|
> | Router   |8888|
> | Broker   |8082|
> | Historical|8083|
> | MiddleManager|8091|

Reviewer 2: ```suggestion
| Service       | Port address |
| ------------- | ------------ |
| Coordinator   | 8081         |
| Overlord      | 8081         |
| Router        | 8888         |
| Broker        | 8082         |
| Historical    | 8083         |
| MiddleManager | 8091         |
```

  ~ under the License.
  -->

This document describes the API endpoints to retrieve service status, cluster information for Apache Druid.

In this document, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://ROUTER_IP:ROUTER_PORT` with `http://localhost:8888`.

## Common

All services support the following endpoints.

You can use each endpoint with the ports for each type of service.  The following table contains port addresses for a local configuration:

> |Service|Port address|
> | ----------- | ----------- |
> | Coordinator|8081|
> | Overlord|8081|
> | Router|8888|
> | Broker|8082|
> | Historical|8083|
> | MiddleManager|8091|

Reviewer 2: ```suggestion
| Service       | Port address |
| ------------- | ------------ |
| Coordinator   | 8081         |
| Overlord      | 8081         |
| Router        | 8888         |
| Broker        | 8082         |
| Historical    | 8083         |
| MiddleManager | 8091         |
```

  ~ under the License.
  -->

This document describes the API endpoints to retrieve service status, cluster information for Apache Druid.

In this document, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://ROUTER_IP:ROUTER_PORT` with `http://localhost:8888`.

## Common

All services support the following endpoints.

You can use each endpoint with the ports for each type of service.  The following table contains port addresses for a local configuration:

|Service|Port address|
| ------ | ------------ |
| Coordinator|8081|
| Overlord|8081|
| Router|8888|
| Broker|8082|
| Historical|8083|
| MiddleManager|8091|

### Get service information

Retrieves the Druid version, loaded extensions, memory used, total memory, and other useful information about the individual service. 

> To retrieve the service information of other services, query the individual port of each service. On a local configuration, refer to this [table](#common) for the port numbers.

Reviewer: ```suggestion
Modify the host and port for the endpoint to match the service to query. Refer to the [default service ports](#common) for the port numbers.
```

            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.MSQServiceClientModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.MSQSqlModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        },
        {
            "name": "org.apache.druid.msq.guice.SqlTaskModule",
            "artifact": "druid-multi-stage-query",
            "version": "26.0.0"
        }
    ],
    "memory": {
        "maxMemory": 268435456,
        "totalMemory": 268435456,
        "freeMemory": 139060688,
        "usedMemory": 129374768,
        "directMemory": 134217728
    }
  }
  ```
</details>

### Get service health

Retrieves the online status of the individual Druid service. It is a simple health check to determine if the service is running and accessible. If online, it will always return a boolean `true` value, indicating that the service can receive API calls. This endpoint is suitable for automated health checks.

> To retrieve the service health of other services, query the individual port of each service. On a local configuration, refer to this [table](#common) for the port numbers.

Reviewer: ```suggestion
Modify the host and port for the endpoint to match the service to query. Refer to the [default service ports](#common) for the port numbers.
```

Additional checks for readiness should use the [Historical segment readiness](#get-segment-readiness) and [Broker query readiness](#get-broker-query-readiness) endpoints. 

#### URL

<code class="getAPI">GET</code> <code>/status/health</code>

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->

<br/>

*Successfully retrieved service health*  

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample request

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->

```shell
curl "http://ROUTER_IP:ROUTER_PORT/status/health"
```

<!--HTTP-->

```http
GET /status/health HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
```

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response

<details>
  <summary>Click to show sample response</summary>

  ```json
  true
  ```

</details>

### Get configuration properties

Retrieves the current configuration properties of the individual service queried. 

> To retrieve the service configuration property of other services, query the individual port of each service. On a local configuration, refer to this [table](#common) for the port numbers.

Reviewer: ```suggestion
Modify the host and port for the endpoint to match the service to query. Refer to the [default service ports](#common) for the port numbers.
```

  ~ under the License.
  -->

This document describes the API endpoints to retrieve service status, cluster information for Apache Druid.

In this document, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port. For example, on the quickstart configuration, replace `http://ROUTER_IP:ROUTER_PORT` with `http://localhost:8888`.

## Common

All services support the following endpoints.

You can use each endpoint with the ports for each type of service.  The following table contains port addresses for a local configuration:

|Service|Port address|

> | ----------- | ----------- |

Reviewer 3: ```suggestion
|--------|------------|
```
